File: LeetCode/857. Minimum Cost to Hire K Workers/857.java
import java.util.Arrays;
import java.util.PriorityQueue;

class Solution {
    public double mincostToHireWorkers(int[] quality, int[] wage, int k) {
        int n = quality.length;
        Worker[] workers = new Worker[n];
        
        for (int i = 0; i < n; i++) {
            workers[i] = new Worker((double) wage[i] / quality[i], quality[i]);
        }
        
        Arrays.sort(workers, (a, b) -> Double.compare(a.ratio, b.ratio));
        
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> Integer.compare(b, a));
        double minWage = Double.POSITIVE_INFINITY;
        int totalQuality = 0;
        
        for (Worker worker : workers) {
            totalQuality += worker.quality;
            maxHeap.offer(worker.quality);
            
            if (maxHeap.size() > k) {
                totalQuality -= maxHeap.poll();
            }
            
            if (maxHeap.size() == k) {
                minWage = Math.min(minWage, totalQuality * worker.ratio);
            }
        }
        
        return minWage;
    }
    
    class Worker {
        double ratio;
        int quality;
        
        Worker(double ratio, int quality) {
            this.ratio = ratio;
            this.quality = quality;
        }
    }
}

File: LeetCode/857. Minimum Cost to Hire K Workers/857.py
import heapq

class Solution(object):
    def mincostToHireWorkers(self, quality, wage, k):
        """
        :type quality: List[int]
        :type wage: List[int]
        :type k: int
        :rtype: float
        """
        n = len(quality)
        workers = sorted([(float(wage[i]) / quality[i], quality[i]) for i in range(n)])
        
        heap = []
        total_quality = 0
        min_wage = float('inf')
        
        for ratio, q in workers:
            total_quality += q
            heapq.heappush(heap, -q)  # Max heap for quality
            
            if len(heap) > k:
                total_quality += heapq.heappop(heap)
            
            if len(heap) == k:
                min_wage = min(min_wage, total_quality * ratio)
        
        return min_wage
